text/0001-null-undefined-usage.md

```diff
@@ -0,0 +1,92 @@
+- Start Date: (2022-12-07)
+- RFC PR: (leave this empty)
+- Issue: (leave this empty)
+
+# Summary
+
+As a contributor to Webstudio codebases we need to agree when to use `undefined` and when to use `null`.
+
+Prior art:
+
+- [Article](https://oleg008.medium.com/what-if-we-stop-using-null-d705302b545e)
+- [PR discussion](https://github.com/sindresorhus/meta/discussions/7)
+- [Presentation by Douglas Clockford](https://www.youtube.com/watch?v=PSGEjv3Tqo0&t=561s)
+- [Tweet from creattor of JavaScript](https://twitter.com/brendaneich/status/617450289889607681) and [this](https://twitter.com/brendaneich/status/330775086208524288)
+- [Internal TypeScript convention for the compiler itself](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined)
+
+# Motivation
+
+Web platform uses both and both have their advantages and downsides, so we want to decide which adavntages we prefer over disadvantages.
+
+Advatages to use `undefined`:
+
+- Default function parameters (JavaScript does not support `null` with default arguments)
+
+  ```js
+  const f = (a = 1) => a;
+  f(); // 1
+  f(undefined); // 1
+  f(null); // null
+  ```
+
+- Destructuring with default values
+
+  ```js
+  const { a = 1 } = { a: null };
+  a; // null
+  const { a = 1 } = {};
+  a; // 1
+  const { a = 1 } = { a: undefined };
+  a; // 1
+  const [a = 1] = [];
+  a; // 1
+  const [a = 1] = [undefined];
+  a; // 1
+  const {
+    a: { b = 1 },
+  } = { a: { b: undefined } };
+  b; // 1
+  ```
+
+- Using null makes TypeScript types more verbose: `type A = {foo?: string | null}` vs `type A = {foo?: string}`.
```

Comment:
Note that if you ever enable [`https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) then `?` won't be adding `| undefined` to your values. 

This option is quite invasive and perhaps you might decide to never turn it on. I recently had some bugs that this option would help me find before shipping... so personally, I think I will be enabling this in my projects. All libraries should enable this anyway - as libraries should be written with the safest options available. Library authors can't decide/assume what kind of options might be enabled by their consumers.

Just saying that with this option, the verbosity wouldn't be as strong. With this option, It's often convenient to add that extra `| undefined` to your types.

Reply:
I won't be able to pass fn(x, undefined) where fn(x: string, b?: string)?

Reply:
This works just fine - this option is related to object properties. See the demo here:
[TS playground](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVNQAoAPALngGcMYtUBzAGngCMB+cqm+gSnIDccWYAFgAUGMREA5FO4TpUpmlCJaIYHNGTCMpjM1iMATwAOCAGI4c8ALzwA3mPjwo5VMgC2zEDCct2lNS0dGIAvmJiYHhU8BgAjOSW1nb2LuRxAAwZ8OGiUagxGABMiVa2DmnwmRlMzOTKIKqo6jmR0RixAMylyRWuVVm15DKt4qJAA)

Reply:
Oh that makes sense now, I am happy to add this option then, btw wanna try add it and see if its hard?

Reply:
I can try to spike it sometime later this week.

Reply:
I want this option too ;) removes ton of | undefined from types and makes things more predictable, I use it in every project too